src/analysis/rule_extraction.jl

```diff
@@ -77,16 +77,97 @@ end
 
 # Arguments
 - `rules` : Vector of Rule objects
 """
-function print_rules(rules::Vector{Rule{Vector{Float64},Vector{Vector{Any}}}})
+function print_rules(rules::Vector{Rule{Vector{Vector},Vector{Float64}}})
     condition(rule) = [c == :L ? "$(c[1]) < $(c[3])" : "$(c[1]) ≥ $(c[3])" for c in rule.condition]
     consequent(rule) = " then $(rule.consequent[1]) else $(rule.consequent[2])\n"
     rule_string(rule) = "if " * join(condition(rule), " ") * consequent(rule)
     print(join([rule_string(rule) for rule in rules]))
 end
+
+
+"""
+    cluster_rules(clusters::Vector{T}, X::DataFrame, outcomes::AbstractMatrix{F}, max_rules::Int64; n_trees::Int64=1000) where {T<:Integer,F<:Real}
+
+Use SIRUS package to extract rules from time series clusters based on some summary metric (default is median)
+
+# Arguments
+- `clusters` - Vector of cluster indexes for each scenario outcome
+- `X::DataFrame` - DataFrame with factors to be used as input by SIRUS
+- `outcomes` - Matrix of some metric over time for all scenarios
+- `max_rules` - Maximum number of rules, to be used as input by SIRUS
+- `n_trees` - Number of trees to be created by SIRUS algorithm
+
+# Returns
+A StableRules object (implemented by SIRUS).
+
+# References
+1. Huijzer, R.
+   rikhuijzer/SIRUS.jl [Computer software].
+   https://doi.org/10.5281/zenodo.7875577
+
+2. Bénard, C., Biau, G., Da Veiga, S., & Scornet, E. 2021.
+   Sirus: Stable and interpretable rule set for classification.
+   Electron. J. Statist. 15 (1) 427 - 505.
+   https://doi.org//10.1214/20-EJS1792
+"""
+function cluster_rules(clusters::Vector{T}, X::DataFrame, outcomes::AbstractMatrix{F},
+    max_rules::T; n_trees::T=1000) where {T<:Int64,F<:Real}
+    # Find robust cluster index
+    rc = robust_cluster(clusters, outcomes)
+    y = clusters .== rc
+
+    # Set seed and Random Number Generator
+    seed = 123
+    _rng(seed::Int=1) = StableRNG(seed)
+
+    # Use SIRUS Stable Rules Classifier model to extract the rules
+    model = StableRulesClassifier(; max_rules=max_rules, n_trees=n_trees)
+    mach = machine(model, X, y)
+    fit!(mach)
+    return rules(mach.fitresult)
+end
+
+"""
+    robust_cluster(clusters::Vector{T}, result_set::ResultSet) where {T<: Integer}
+
+Find most robust cluster.
+
+# Arguments
+- `clusters` - Vector with outcomes cluster indexes
+- `outcomes` - AbstractMatrix of outcomes factors
+
+# Returns
+Index of the most robust cluster
+"""
+function robust_cluster(clusters::Vector{T}, outcomes::AbstractMatrix{F}) where {T<:Int64,F<:Real}
+    clusters_statistics::Vector{Float64} = []
+
+    max_outcome = maximum(outcomes)
+    for cluster in unique(clusters)
+        normalized_outcomed = outcomes[:, clusters.==cluster] ./ max_outcome
+        statistic = median(temporal_variability(normalized_outcomed; w=[0.9, 0.1]))
+        push!(clusters_statistics, statistic)
+    end
+
+    # Find index of the cluster with maximum value of summary statistic
+    return argmax(clusters_statistics)
+end
```

Comment:
Add kwarg to allow any function to be passed in to be used in place of `temporal_variability`. 

There are many (meta-)metrics that could be adopted here, so it would good to have the flexibility.

Reply:
Why `kwargs` and not a variable called `metric`?

Reply:
I was assuming you want a default metric...

Reply:
What do you think changing this function's name to `target_cluster` instead?

Reply:
About the `kwargs` I was thinking about having a `metric` argument setting `temporal_variability` as the default value. But I think is "a better approach" to set this default value inside the function, in this case?

Reply:
> What do you think changing this function's name to `target_cluster` instead?

Could be. Another option is to use a term borrowed from statistics: `behavioral`

The cluster with the desirable properties are `behavioral` and those that are not are `non-behavioral`.
So `behavioral_cluster` or `identify_behavioral` ?

Reply:
> About the kwargs I was thinking about having a metric argument setting temporal_variability as the default value. But I think is "a better approach" to set this default value inside the function, in this case?

Could you elaborate on this? Is this for flexibility in some way?

Reply:
Maybe if my OO side speaking here, but I prefer `behavioral_cluster` - to me what, in this case what is important is not what this function does, but the information it returns. So I prefer a name the describes what is returns than what is does. But I'm happy either way (:

Reply:
Haha, it's always a contentious debate. Some people say functions/methods should relate to what they do ("functions are actions"), with documentation on what/why, for example: `Dog.speak()`, `Printer.print()`

I'm happy with `behavioral_cluster` or `target_cluster`, I'll leave it up to you.

Reply:
> Could you elaborate on this? Is this for flexibility in some way?

Actually I think I prefer you suggestion. But initially I thought of doing something like:

```
function behavioral_cluster(clusters::Vector{T}, outcomes::AbstractMatrix{F}; metric=temporal_variability)
```

And my question was just why one option is better than the other?

Reply:
> > Could you elaborate on this? Is this for flexibility in some way?
> 
> Actually I think I prefer you suggestion. But initially I thought of doing something like:
> 
> ```
> function behavioral_cluster(clusters::Vector{T}, outcomes::AbstractMatrix{F}; metric=temporal_variability)
> ```

This is what I was suggesting actually. When you said:

`But I think is "a better approach" to set this default value inside the function, in this case?`

By "inside the function" I was imagining you meant something like:

```julia
function behavioral_cluster(...; metric=nothing)
    if isnothing(metric)
        metric = temporal_variability
    end
end
```

But (I guess) you meant inside the function signature rather than inside its definition.

Reply:
What I meant is what I just pushed:

```
function target_cluster(clusters::Vector{T}, outcomes::AbstractMatrix{F}; kwargs...)
    metric = haskey(kwargs, :metric) ? kwargs[:metric] : temporal_variability

    ...
```

Reply:
Ah, well in that case I would say if you always expect a `metric` keyword arg, why not just specify it?

The only time I would adopt that pattern is if I expect to need to pass multiple (but separate) kwargs to other functions or if this function is a barrier function that does some preprocessing for some other functions (for example, how `Metrics` are implemented in ADRIA or the visualizations).

Here, `target_cluster` is a single function, with no other kwargs expected.